Fail to require (in cjs) CSS modules from an npm package

### Describe the bug

The background is I have some existing packages whose generated code is in CommonJS + CSS modules. E.g.

```js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.circularFont = void 0;
const circular_font_module_scss_1 = __importDefault(require("./circular-font.module.scss"));
exports.circularFont = circular_font_module_scss_1.default.circularFont;
//# sourceMappingURL=index.js.map
```

This file uses `require()` to load CSS Modules.

In Vite, it fails to be loaded, however, an `import` works.

### Reproduction

https://github.com/Jinjiang/reproductions/tree/vite-cssmodules-20230523

### Steps to reproduce

```
pnpm install
pnpm dev
```

Then open browser, the console panel will show up:

```js
import a css module {foo: '_foo_iyjzy_1', bar: '_bar_iyjzy_5'} // works
require a css module {__esModule: true} // not
{foo: '_foo_iyjzy_1', bar: '_bar_iyjzy_5'} // works
undefined // not
{foo: '_foo_1clpo_1', bar: '_bar_1clpo_5'} // works
{foo: '_foo_iyjzy_1', bar: '_bar_iyjzy_5'} // works
```

### System Info

```shell
$ npx envinfo --system --npmPackages '{vite,@vitejs/*}' --binaries --browsers

  System:
    OS: macOS 13.4
    CPU: (8) arm64 Apple M1
    Memory: 69.14 MB / 8.00 GB
    Shell: 5.9 - /bin/zsh
  Binaries:
    Node: 16.15.0 - ~/.asdf/installs/nodejs/16.15.0/bin/node
    Yarn: 1.22.19 - ~/.asdf/shims/yarn
    npm: 8.5.5 - ~/.asdf/plugins/nodejs/shims/npm
  Browsers:
    Safari: 16.5
  npmPackages:
    vite: ^4.3.8 => 4.3.8
```


### Used Package Manager

pnpm

### Logs

_No response_

### Validations

- [X] Follow our [Code of Conduct](https://github.com/vitejs/vite/blob/main/CODE_OF_CONDUCT.md)
- [X] Read the [Contributing Guidelines](https://github.com/vitejs/vite/blob/main/CONTRIBUTING.md).
- [X] Read the [docs](https://vitejs.dev/guide).
- [X] Check that there isn't [already an issue](https://github.com/vitejs/vite/issues) that reports the same bug to avoid creating a duplicate.
- [X] Make sure this is a Vite issue and not a framework-specific issue. For example, if it's a Vue SFC related bug, it should likely be reported to [vuejs/core](https://github.com/vuejs/core) instead.
- [X] Check that this is a concrete bug. For Q&A open a [GitHub Discussion](https://github.com/vitejs/vite/discussions) or join our [Discord Chat Server](https://chat.vitejs.dev/).
- [X] The provided reproduction is a [minimal reproducible example](https://stackoverflow.com/help/minimal-reproducible-example) of the bug.

@haijie-x thanks for the information. I tried to install this package and added vite config file you provided. Then I got another error:

```
Error:   Failed to scan for dependencies from entries:
  C:/Users/zhaoj/code/reproductions/index.html

  X [ERROR] The JSX syntax extension is not currently enabled

    html:C:/Users/zhaoj/code/reproductions/index.html:1:0:
      1 │ <script type="module" src="index.js"></script>
        ╵ ^

  The esbuild loader for this file is currently set to "js" but it must be set to "jsx" to be able to parse JSX syntax. You can use "loader: { '.js': 'jsx' }" to do that.


    at failureErrorWithLog (C:\Users\zhaoj\code\reproductions\node_modules\.pnpm\esbuild@0.17.19\node_modules\esbuild\lib\main.js:1636:15)
    at C:\Users\zhaoj\code\reproductions\node_modules\.pnpm\esbuild@0.17.19\node_modules\esbuild\lib\main.js:1048:25
    at runOnEndCallbacks (C:\Users\zhaoj\code\reproductions\node_modules\.pnpm\esbuild@0.17.19\node_modules\esbuild\lib\main.js:1471:45)
    at buildResponseToResult (C:\Users\zhaoj\code\reproductions\node_modules\.pnpm\esbuild@0.17.19\node_modules\esbuild\lib\main.js:1046:7)
    at C:\Users\zhaoj\code\reproductions\node_modules\.pnpm\esbuild@0.17.19\node_modules\esbuild\lib\main.js:1058:9
    at new Promise (<anonymous>)
    at requestCallbacks.on-end (C:\Users\zhaoj\code\reproductions\node_modules\.pnpm\esbuild@0.17.19\node_modules\esbuild\lib\main.js:1057:54)
    at handleRequest (C:\Users\zhaoj\code\reproductions\node_modules\.pnpm\esbuild@0.17.19\node_modules\esbuild\lib\main.js:723:19)
    at handleIncomingPacket (C:\Users\zhaoj\code\reproductions\node_modules\.pnpm\esbuild@0.17.19\node_modules\esbuild\lib\main.js:745:7)
    at Socket.readFromStdout (C:\Users\zhaoj\code\reproductions\node_modules\.pnpm\esbuild@0.17.19\node_modules\esbuild\lib\main.js:673:7)
```

Is there anything I missed? Or do you have any idea how to make it works?

Thanks.

--- a/packages/vite/src/node/optimizer/esbuildDepPlugin.ts
+++ b/packages/vite/src/node/optimizer/esbuildDepPlugin.ts
@@ -1,6 +1,6 @@
 import path from 'node:path'
 import type { ImportKind, Plugin } from 'esbuild'
-import { CSS_LANGS_RE, KNOWN_ASSET_TYPES } from '../constants'
+import { KNOWN_ASSET_TYPES } from '../constants'
 import { getDepOptimizationConfig } from '..'
 import type { PackageCache, ResolvedConfig } from '..'
 import {
@@ -12,6 +12,7 @@ import {
   normalizePath,
 } from '../utils'
 import { browserExternalId, optionalPeerDepId } from '../plugins/resolve'
+import { isCSSRequest, isModuleCSSRequest } from '../plugins/css'
 
 const externalWithConversionNamespace =
   'vite:dep-pre-bundle:external-conversion'
@@ -172,10 +173,11 @@ export function esbuildDepPlugin(
           // import itself with prefix (this is the actual part of require-import conversion)
           const modulePath = `"${convertedExternalPrefix}${args.path}"`
           return {
-            contents: CSS_LANGS_RE.test(args.path)
-              ? `import ${modulePath};`
-              : `export { default } from ${modulePath};` +
-                `export * from ${modulePath};`,
+            contents:
+              isCSSRequest(args.path) && !isModuleCSSRequest(args.path)
+                ? `import ${modulePath};`
+                : `export { default } from ${modulePath};` +
+                  `export * from ${modulePath};`,
             loader: 'js',
           }
         },

--- a/playground/optimize-deps/__tests__/optimize-deps.spec.ts
+++ b/playground/optimize-deps/__tests__/optimize-deps.spec.ts
@@ -215,6 +215,7 @@ test('pre bundle css require', async () => {
   }
 
   expect(await getColor('.css-require')).toBe('red')
+  expect(await getColor('.css-module-require')).toBe('red')
 })
 
 test.runIf(isBuild)('no missing deps during build', async () => {

--- a/playground/optimize-deps/dep-css-require/mod.cjs
+++ b/playground/optimize-deps/dep-css-require/mod.cjs
@@ -1,0 +1,2 @@
+const style = require('./mod.module.css')
+module.exports = style

--- a/playground/optimize-deps/dep-css-require/mod.module.css
+++ b/playground/optimize-deps/dep-css-require/mod.module.css
@@ -1,0 +1,3 @@
+.cssModuleRequire {
+  color: red;
+}

--- a/playground/optimize-deps/index.html
+++ b/playground/optimize-deps/index.html
@@ -99,6 +99,9 @@
 <h2>Pre bundle css require</h2>
 <div class="css-require">css require</div>
 
+<h2>Pre bundle css modules require</h2>
+<div class="css-module-require">This should be red</div>
+
 <script>
   function text(el, text) {
     document.querySelector(el).textContent = text
@@ -127,6 +130,10 @@
 
   import '@vitejs/test-dep-cjs-with-assets'
   import '@vitejs/test-dep-css-require'
+  import cssModuleRequire from '@vitejs/test-dep-css-require/mod.cjs'
+  document
+    .querySelector('.css-module-require')
+    .classList.add(cssModuleRequire.cssModuleRequire)
 
   import { env } from '@vitejs/test-dep-node-env'
   text('.node-env', env)

